ADNI Data Analysis
========================================================
This code caclculates the slope of the PET volumes from 4 visits and explores the relationship between this slope and the probabilistic grey matter segmentation at baseline. 

Elizabeth Sweeney
June 15, 2014


Load required packages 
```{r}
library(oro.nifti)
library(RColorBrewer)
library(Matrix)
library(nplargedb)
library(nplargela)
library(dplyr)
```

Set working directory 
```{r}
setwd('/dexter/disk1/smart/AD/ADNI/Greven_ADNI')
source("nplarge-regression-utils.R")
## setwd('/Users/elizabethsweeney/Desktop/Greven_ADNI')
```

Grab the subject ids 
```{r}
fileDir <- '/dexter/disk1/smart/AD/ADNI/Greven_ADNI'
files <- dir(fileDir, pattern = "*.nii.gz", full.names = TRUE)
subject.ids <- unlist(lapply(files, function(x) tail(unlist(strsplit(x, '/')), n=1)))
subject.ids <- unique(unlist(lapply(subject.ids, function(x) substr(x, 1, 10))))
```


Read in PET volumes and calculate the slope at each voxel.  The slope volume is then saved. 

```{r}
times <- matrix(c(0,  12, 36, 48),  4, 1)
intercept <- matrix(c(1,1,1,1), 4, 1) 
design <- cbind(intercept, times)
almost.beta <- solve(t(design)%*%design) %*% t(design)

for(i in 1:length(subject.ids)){
  pet1 <- readNIfTI(paste0(subject.ids[i], '_PET_coreg_to_MRI_MNI_1.nii.gz'), reorient = FALSE)[,,]
  pet2 <- readNIfTI(paste0(subject.ids[i], '_PET_coreg_to_MRI_MNI_2.nii.gz'), reorient = FALSE)[,,]
  pet3 <- readNIfTI(paste0(subject.ids[i], '_PET_coreg_to_MRI_MNI_3.nii.gz'), reorient = FALSE)[,,]
	pet4 <- readNIfTI(paste0(subject.ids[i], '_PET_coreg_to_MRI_MNI_4.nii.gz'), reorient = FALSE)[,,]

pet.image <- cbind(c(pet1), c(pet2), c(pet3), c(pet4))
pet.slope <- almost.beta%*%t(pet.image)
pet.slope.volume <- array(pet.slope[2,], dim = dim(pet1))

y.var <- apply(pet.image, 1, function(x) sd(x))
t.stat <- (4 - 2)* pet.slope[2,]/(solve(t(times)%*%times)*y.var) 
p.value <- dt(t.stat, df = 2)
p.value <- array(p.value , dim = dim(pet1))

writeNIfTI(p.value, paste0(subject.ids[i], '_pvalues'))

correction.size <- mean( c(sum(pet1 > .01, na.rm = TRUE) , sum(pet2 > .01, na.rm = TRUE), sum(pet3 > .01, na.rm = TRUE), sum(pet4 > .01, na.rm = TRUE)))

p.value.bonferoni.mask <- array(0, dim = dim(p.value))
p.value[is.na(p.value)] <- 10
p.value.bonferoni.mask[p.value  < .05/correction.size] <- 1
pet4[is.na(pet4)] <- 0
p.value.bonferoni.mask[pet4 < .01] <- 0


writeNIfTI(p.value.bonferoni.mask, paste0(subject.ids[i], '_bonfmask'))


writeNIfTI(pet.slope.volume, paste0(subject.ids[i], '_PETslope'))
print(i)
}
```

Explore relationship between PET slope and probabilistic grey matter segmentation in different ROIs

Hippocampus
```{r}
grey.matter <- c()
pet.slopes <- c()

ROI <- readNIfTI('/dexter/disk1/smart/AD/ADNI/Greven_ADNI/AAL_masks/Hippocampus.nii.gz', reorient = FALSE)[,,]

for(i in 1:length(subject.ids)){
  PET <- readNIfTI(paste0(subject.ids[i], '_PETslope'), reorient = FALSE)[,,]
  GM <- readNIfTI(paste0(subject.ids[i], '_MRI_segm_MNI_1.nii.gz'), reorient = FALSE)[,,]
  grey.matter <- c(grey.matter, GM[ROI == 1]) 
  pet.slopes <- c(pet.slopes, PET[ROI == 1])
  print(i) 
  }

pdf('Hippo.pdf')
down.samp <- sample(1:length(grey.matter), 5000, replace = FALSE)
smooth <- loess.smooth(grey.matter, pet.slopes) 
plot(pet.slopes[down.samp] ~ grey.matter[down.samp], pch = 20, main = 'Hippocampus', ylab = 'PET slope', xlab = c('Baseline Grey Matter'))
lines(smooth, lwd = 3, col = 'red')
dev.off()

pdf('SubjectsHippo.pdf')
down.samp <- sample(1:length(grey.matter), 5000, replace = FALSE)
plot(pet.slopes[down.samp] ~ grey.matter[down.samp], pch = 20, main = 'Hippocampus by Subject', ylab = 'PET slope', xlab = c('Baseline Grey Matter'))


colors.mine <- rep(brewer.pal(9, 'Set1'), 12)

for(i in 1:length(subject.ids)){
  PET <- readNIfTI(paste0(subject.ids[i], '_PETslope'), reorient = FALSE)[,,]
  GM <- readNIfTI(paste0(subject.ids[i], '_MRI_segm_MNI_1.nii.gz'), reorient =   FALSE)[,,] 
  subj.grey.matter <- GM[ROI == 1]
  subj.pet.slopes <- PET[ROI == 1]
  smooth <- loess.smooth(subj.grey.matter, subj.pet.slopes) 
  lines(smooth, lwd = 1, col = colors.mine[i])
  print(i)
}
dev.off()
```

Investigate relationship in the Hippocampus for only statistically significant slopes. (Used a Bonferoni correction over the entire brain, might consider just doing for the Hippocampus as next step)

```{r}

grey.matter.bonf <- c()
pet.slopes.bonf <- c()

ROI <- readNIfTI('/dexter/disk1/smart/AD/ADNI/Greven_ADNI/AAL_masks/Hippocampus.nii.gz', reorient = FALSE)[,,]

for(i in 1:length(subject.ids)){
  PET <- readNIfTI(paste0(subject.ids[i], '_PETslope'), reorient = FALSE)[,,]
  GM <- readNIfTI(paste0(subject.ids[i], '_MRI_segm_MNI_1.nii.gz'), reorient = FALSE)[,,]
  bonf.correct <- readNIfTI(paste0(subject.ids[i], '_bonfmask'), reorient = FALSE)[,,]
  bonf.roi <- bonf.correct*ROI
  grey.matter.bonf <- c(grey.matter.bonf, GM[bonf.roi == 1]) 
  pet.slopes.bonf <- c(pet.slopes.bonf, PET[bonf.roi == 1])
  print(i) 
  }

pdf('HippoBonf.pdf')
down.samp <- sample(1:length(grey.matter.bonf), 5000, replace = FALSE)
smooth <- loess.smooth(grey.matter.bonf, pet.slopes.bonf) 
plot(pet.slopes.bonf[down.samp] ~ grey.matter.bonf[down.samp], pch = 20, main = 'Hippocampus Bonferoni', ylab = 'PET slope', xlab = c('Baseline Grey Matter'))
lines(smooth, lwd = 3, col = 'red')
dev.off()


```


Make PLOT of the PET slope volume for one subject 
```{r}

PET <- readNIfTI(paste0(subject.ids[1], '_PETslope'), reorient = FALSE)[,,]
PET[is.na(PET)] <- 0

pdf('petslopes.pdf')
orthographic(PET, col = hotmetal())
dev.off()  

mean.pet <- array(0, dim = dim(PET))
for(i in 1:length(subject.ids)){
  PET <- readNIfTI(paste0(subject.ids[i], '_PETslope'), reorient = FALSE)[,,]
  PET[is.na(PET)] <- 0
  mean.pet <- ((i-1)*mean.pet + PET)/i
  }

var.pet <- array(0, dim = dim(PET))
for(i in 1:length(subject.ids)){
  PET <- readNIfTI(paste0(subject.ids[i], '_PETslope'), reorient = FALSE)[,,]
  PET[is.na(PET)] <- 0
  var.pet <- ((i-1)*var.pet + (PET - mean.pet)^2)/i
  }

setwd('/dexter/disk1/smart/AD/ADNI/Greven_ADNI/Output')

writeNIfTI(mean.pet,'meanPETslope')
writeNIfTI(var.pet,'varPETslope')

```

Read in the metadata.  Codebook for the metadata: 

DXCHANGE: diagnosis of the subject at baseline

ADAS11: Alzheimer's disease assessment scale, cognitive subscale (sometimes known as ADAS-Cog) http://www.ncbi.nlm.nih.gov/pubmed/15592137.
It is a neurocognitive test.

APGEN1 and APGEN2: genotype results for APOE variant in 1st and 2nd allelle. A subject is considered a carrier if he/she is APOE-epsilon 4 positive (ie.APGEN1 == 4 or APGEN2 == 4). See reviews for more details on the relevance of APOE.

av: AV45-PET standard uptake value ratio (SUVR) value for the whole brain; for details and methods see ADNI website.

pib: PiB-PET SUVR value for the whole brain; for details and methods see ADNI website.

ABETA142: measurements of cerebro-spinal fluid (CSF) concentrations of Amyloid-beta (Abeta), in pg/mL. Values measured at baseline.

VISCODE.pib: visit for which the PiB-PET scan was performed. Code is bl = baseline, m06 = month 6, m12 = month 12, etc.

VISCODE.av: same for the AV45-PET scan

PTAU181P: measurements of CSF concentrations of phosphorilated-tau protien, in pg/mL. Values measured at baseline.

PTGENDER: subject gender (1 = Male, 2 = Female)

PTEDUCAT: years of education

Ab.status: a patient is considered Abeta positive (Ab.status = 1) if ABETA142 < 192 pg/mL and negative otherwise.

```{r}
meta.data <- read.table('Patient_metadata_3yrs_with_fw-ups.csv', header = TRUE, sep = ',')
baseline.meta.data <- meta.data[meta.data$VISCODE == 'bl',]
baseline.subjects <- merge(data.frame(Subject = c(subject.ids)), baseline.meta.data)
baseline.subjects$MCI <- rep(0, length(baseline.subjects$DXCHANGE))
baseline.subjects$MCI[baseline.subjects$DXCHANGE == 'Stable: MCI'] <- 1

summary(baseline.subjects)
```

Some plots of the baseline data 
```{r}
table(baseline.subjects$APGEN1,baseline.subjects$APGEN2)

##     2  3  4
##  2  1  6  3
##  3  0 46 30
##  4  0  0  8

pdf('Abeta.pdf')
hist(baseline.subjects$ABETA142, main = 'Amyloid-beta meaurements from the CSF', breaks = 25)
dev.off()

pdf('age.pdf')
hist(baseline.subjects$AGE, main= 'Age', breaks = 25)
dev.off()

pdf('tau.pdf')
hist(baseline.subjects$PTAU181P, main = 'CSF concentrations of phosphorilated-tau protien', breaks = 25)
dev.off()

pdf('logtau.pdf')
hist(log(baseline.subjects$PTAU181P), main = 'Log CSF concentrations of phosphorilated-tau protien', breaks = 25)
dev.off()

table(baseline.subjects$PTGENDER)

## Male = 1, Female = 2
## 1  2 
##63 31

pdf('education.pdf')
hist(baseline.subjects$PTEDUCAT, main = 'Years of Education')
dev.off()

table(baseline.subjects$DXCHANGE)
          Stable: MCI 
               53 
           Stable: NL 
               41 


pdf('ADAS11.pdf')
hist(baseline.subjects$ADAS11, main = 'Alzheimers disease assessment scale')
dev.off()

```

Create an image to use for the regression (will be getting an ROI from the collaborators, for now just use an arbitrary rectangle...from inside the Hippocampus mask...?)

```{r} 
Hippo.ROI <- readNIfTI('/dexter/disk1/smart/AD/ADNI/Greven_ADNI/AAL_masks/Hippocampus.nii.gz', reorient = FALSE)[,,]
sum(Hippo.ROI[36:40, 66:70, 41:45])/c(5*5*5)
##0.832
ROI.mask <- array(0, dim = dim(Hippo.ROI))
ROI.mask[36:40, 66:70, 41:45] <- 1 
```

Read in the PET slope data
```{r}
## enter information about the number of subjects and the dimmensions of the ROI
n <-  94
dim_roi <- c(5, 5, 5)
n_dims <- length(dim_roi) 
dims <- lapply(dim_roi, function(x) seq_len(x))
names(dims) <- paste0("dim", 1:n_dims)

pet.slope.outcome <- c()
for(i in 1:length(subject.ids)){
   PET <- readNIfTI(paste0(subject.ids[i], '_PETslope'), reorient = FALSE)[,,]
   pet.slope.outcome <- c(pet.slope.outcome, PET[ROI.mask == 1])
   print(i)
}

y <- array(pet.slope.outcome, c(dim_roi, n))


y_cube <- tbl_cube(dimensions=append(dims, list(Subject=subject.ids)), 
                   measures = list(y=y))

```
Preparing this data for use with the nplarge packages 

```{r}

## make an array the size of the ROI for each covariate of interest (we will begin by focusing on age, gender, education, and MCI)
keep <- c("AGE", "PTGENDER", "MCI", "PTEDUCAT", "Subject")
data <- baseline.subjects[keep]
##age.array <- aperm(array(baseline.subjects$AGE, dim=c(n, dim_roi)), c(2:(n_dims+1), 1))
##gender.array <- aperm(array(baseline.subjects$PTGENDER, dim=c(n, dim_roi)), c(2:(n_dims+1), 1))
##education.array <- aperm(array(baseline.subjects$PTEDUCAT, dim=c(n, dim_roi)), c(2:(n_dims+1), 1))
##MCI.array <- aperm(array(baseline.subjects$MCI, dim=c(n, dim_roi)), c(2:(n_dims+1), 1))

##Store the covariate arrays in a data cube tbl
##cov_cube <- tbl_cube(dimensions=append(dims, list(subject=1:n)), 
##                   measures = list(MCI = MCI.array, EDU = education.array, 
##                                   GEN = gender.array, AGE = age.array))

denv <- dataenv(data,  dims="Subject")
add_dims(data=y_cube, dataenv=denv)

##Creat a compressed data representaion of data cube tbl
## denv <- dataenv(cov_cube)

## Create intercept term for subjects
const_subj <- const(n, denv)

##Create intercept term for the dimmensions of the ROI
const_v <- const(prod(dim_roi), denv)

##Create bspline basis
bs_dims <- lapply(paste0("dim", 1:n_dims), 
                  function(dim) bspline(dim, denv, df=6, constraint="none"))
bs_v <- do.call(kron, bs_dims)

##Create intercept array 
intercept <- kron(const_v, const_subj)

##create bspline basis for age, mci, education, and gender
bs_age <- kron(bs_v, const(n, denv, by=baseline.subjects$AGE))
bs_mci <- kron(bs_v, const(n, denv, by=baseline.subjects$MCI))
bs_edu <- kron(bs_v, const(n, denv, by=baseline.subjects$PTEDUCAT))
bs_gen <- kron(bs_v, const(n, denv, by=baseline.subjects$PTGENDER))


pred <- concat(intercept, bs_age, bs_mci, bs_edu, bs_gen)
resp <- pet.slope.outcome 

m <- nplarge_lm_fit(pred, resp)

##get the fitted values 
y_hat <- array(m$fitted.values, c(dim_roi, n))

```

Visualization of results 
```{r}

subjects <- sample(1:n, 4)
slices <- sample(1:dim_roi[3], 4)
clrs <- heat.colors(24)
brks <- seq(min(resp), max(resp), l=25) 

pdf('scalar_reg_est.pdf')
for(i in 1:4){
  par(mfrow = c(1,2))
        image(y_cube$mets$y[, , slices[i], subjects[i]], col=clrs, breaks=brks,
              main="y")
        image(y_hat[, , slices[i], subjects[i]], col=clrs, breaks=brks, 
              main="est")    
    }
dev.off()

# visualize beta-estimates:
## intercept 
m$coef[1]
coef_age <- m$coef[2 : cumsum(pred$c)[2]]
coef_mci <- m$coef[c(cumsum(pred$c)[2] + 1) : cumsum(pred$c)[3]]
coef_edu <- m$coef[c(cumsum(pred$c)[3] + 1) : cumsum(pred$c)[4]]
coef_gen <- m$coef[c(cumsum(pred$c)[4] + 1) : cumsum(pred$c)[5]]

beta_age_hat <- array(pred$terms[[2]]$getM(use_by=FALSE) %*% coef_age,
                    dim=c(dim_roi, n))
beta_mci_hat <- array(pred$terms[[3]]$getM(use_by=FALSE) %*% coef_mci,
                    dim=c(dim_roi, n))
beta_edu_hat <- array(pred$terms[[3]]$getM(use_by=FALSE) %*% coef_edu,
                    dim=c(dim_roi, n))
beta_gen_hat <- array(pred$terms[[3]]$getM(use_by=FALSE) %*% coef_gen,
                    dim=c(dim_roi, n))


pdf('beta_age.pdf')
  clrs <- heat.colors(24)
    slices <- round(seq(1, dim_roi[3]))
   par(mfrow = c(2,3))

    brks <- seq(min(beta_age_hat), max(beta_age_hat), l=25) 
    for(slice in slices){
        image(beta_age_hat[,,slice,1], main=bquote("beta_age, slice "~.(slice)),
              col=clrs, breaks=brks)
    }
dev.off()

pdf('beta_mci.pdf')
  clrs <- heat.colors(24)
    slices <- round(seq(1, dim_roi[3]))
   par(mfrow = c(2,3))

    brks <- seq(min(beta_mci_hat), max(beta_mci_hat), l=25) 
    for(slice in slices){
        image(beta_mci_hat[,,slice,1], main=bquote("beta_mci, slice "~.(slice)),
              col=clrs, breaks=brks)
    }
dev.off()


pdf('beta_edu.pdf')
  clrs <- heat.colors(24)
    slices <- round(seq(1, dim_roi[3]))
   par(mfrow = c(2,3))

    brks <- seq(min(beta_edu_hat), max(beta_edu_hat), l=25) 
    for(slice in slices){
        image(beta_edu_hat[,,slice,1], main=bquote("beta_edu, slice "~.(slice)),
              col=clrs, breaks=brks)
    }
dev.off()


pdf('beta_gen.pdf')
  clrs <- heat.colors(24)
    slices <- round(seq(1, dim_roi[3]))
   par(mfrow = c(2,3))

    brks <- seq(min(beta_gen_hat), max(beta_gen_hat), l=25) 
    for(slice in slices){
        image(beta_gen_hat[,,slice,1], main=bquote("beta_gen, slice "~.(slice)),
              col=clrs, breaks=brks)
    }
dev.off()
```

For a penalized fit...
```{r}

##add a difference penalty to all covariates (note that we do not add the penalty to the intercept)

    pred$terms[[2]]$pen <- do.call(make_kron_sum, 
                                   lapply(pred$terms[[2]]$terms[1:n_dims], 
                                          make_diffpen))
    pred$terms[[3]]$pen <- do.call(make_kron_sum, 
                                   lapply(pred$terms[[3]]$terms[1:n_dims], 
                                          make_diffpen))
    pred$terms[[4]]$pen <- do.call(make_kron_sum, 
                                   lapply(pred$terms[[3]]$terms[1:n_dims], 
                                          make_diffpen))
    pred$terms[[5]]$pen <- do.call(make_kron_sum, 
                                   lapply(pred$terms[[3]]$terms[1:n_dims], 
                                          make_diffpen))

test_opt <- nplarge_pen_fit(x=pred, y=resp, optimizer="optim-BFGS",
                                     optimizer.control=list(trace=3))
                            
```
